src/led/libled.h

```diff
@@ -0,0 +1,492 @@
+ *  - You must set a valid open file descriptor before any log messages will be written.
+ */
+void LED_SYM_PUBLIC led_log_level_set(struct led_ctx *ctx, enum led_log_level_enum level);
+
+/**
+ * @brief Add a device path to an include list.  The library will then only consider this device
+ * for API calls.  Function can be repeated, to add to an internal list.
+ * 
+ * @param[in]	ctx	Library context
+ * @param]in]	path	Device path in include
+ * @return led_status_t	LED_STATUS_SUCCESS on sucess, else error reason.
+ * 
+ * Note: Mutually exclusive with exclude, include takes priority.
+ */
+led_status_t LED_SYM_PUBLIC led_device_include_add(struct led_ctx *ctx, const char *path);
+
+/**
+ * @brief Add a device path to an exclude list.  The library will then exclude this device
+ * from API calls.  Function can be repeated, to add to an internal list.
+ * 
+ * @param[in]	ctx	Library context 
+ * @param[in]	path	Device path to exclude 
+ * @return led_status_t LED_STATUS_SUCCESS on sucess, else error reason.
+ */
+led_status_t LED_SYM_PUBLIC led_device_exclude_add(struct led_ctx *ctx, const char *path);
+
+/**
+ * @brief Sets library blink behavior.
+ * 
+ * @param[in]	ctx	Library context 
+ * @param[in]	migration 
+ * @param[in]	init 
+ * @param[in]	all 
+ * @param[in]	raid_members 
+ * @return led_status_t LED_STATUS_SUCCESS on sucess, else error reason.
+ */
+led_status_t LED_SYM_PUBLIC led_device_blink_behavior_set(struct led_ctx *ctx, int migration,
+								int init, int all, int raid_members);
+
+/**
+ * @brief Instructs the library to scan system hardware for block devices with LED support.  This needs
+ * to be called before any other library functions can be utilized.  Can be called to update any changes
+ * of state to the hardware.
+ * 
+ * @param[in]	ctx	Library context 
+ * @return led_status_t LED_STATUS_SUCCESS on sucess, else error reason.
+ */
+led_status_t LED_SYM_PUBLIC led_scan(struct led_ctx *ctx);
+
+/**
+ * @brief Used to lookup a block device node to name used by library
+ * 
+ * @param[in] 	name			Device node to lookup
+ * @param[out] 	normalized_name		Normalized device name 
+ * @return led_status_t LED_STATUS_SUCCESS on sucess, else error reason.
+ * 
+ * Note: both parameters are expected to have a size of PATH_MAX
+ */
+led_status_t LED_SYM_PUBLIC led_device_name_lookup(const char *name, char *normalized_name);
+
+/**
+ * @brief Given a block device path, returns if it has LED hardware support via library
+ * 
+ * @param[in]	ctx	Library context 
+ * @param[in]	path	Device path 
+ * @return led_status_t LED_STATUS_SUCCESS to indicate support, else error reason.
+ */
+led_status_t LED_SYM_PUBLIC led_support(struct led_ctx *ctx, const char *path);
+
+/**
+ * @brief Turns off all LEDs for all support hardware for the system
+ * 
+ * @param[in] ctx	Library context
+ * 
+ * Note: Needs to be followed with led_flush() 
+ */
+void LED_SYM_PUBLIC led_off_all(struct led_ctx *ctx);
```

Comment:
I think that something like that should not be defined. The first is using default controller (it is not what user may expect).
The second objection is that API should allow user to define what he want to do, not to do job for him. Caller must be aware of controller he is using.

Reply:
From what I understand, this iterates over all the controllers, not just a default, but maybe I'm incorrect?  The API entry points which only specify the device path, the user has no knowledge of the controller that is used, simply device & pattern.  Like I mentioned before, the get/set slot by controller is not available for all supported LED hardware.

Note: This was added for ledctl.

Reply:
> From what I understand, this iterates over all the controllers, not just a default, but maybe I'm incorrect?

The device could have multiple controllers supported (for example NPEM and VMD). There is a mechanism in ledmon for choosing the best one- in implementation you are using it because `send_fn` it set to the best match.
I other comment I asked you to export "best controller match function" for particular device.

> The API entry points which only specify the device path, the user has no knowledge of the controller that is used, simply device & pattern. Like I mentioned before, the get/set slot by controller is not available for all supported LED hardware

And this is why library use should not use it. We cannot predict the result at all.
Perhaps other, independent disk may have some led management and it's status will be cleared by this call. Such implementation should be done in application, not in library.

In my conception, I tract ledmon and ledctl as application based on led API. I understand that It is extremally hard to divide this code now and appreciate your input.

> Note: This was added for ledctl.

Ok, for now we can left it this way. It is a know issue that ledctl is not a just simpler setter but it does more in background.
I think that library doesn't need block list at all but we need to remove this dependency from ledctl first- it is separate problem.

Reply:
> > The API entry points which only specify the device path, the user has no knowledge of the controller that is used, simply device & pattern. Like I mentioned before, the get/set slot by controller is not available for all supported LED hardware
> 
> And this is why library use should not use it. We cannot predict the result at all. Perhaps other, independent disk may have some led management and it's status will be cleared by this call. Such implementation should be done in application, not in library.

Does this imply the existing ledctl is flawed?

Reply:
Ledctl does much more that it should but there is an explanation.
In the past main use case for customers was to blink locate for particular drive. We also wanted mechanism to clear locate mark from other drives (if they are marked with locate already).
For that reason with default settings, it does one `_detemine()` loop and updates leds on every supported drive to _determined (compatibility with ledmon). 

Now things changed, ledctl is sometimes used by customers to define they own blinking apps, we added `--listed-only` to avoid changes on other drives. 
We planned to add real communication between ledctl and ledmon, to make ledmon as primary led manager (if working) and impose every led change request from ledctl by ledmon. With this solution, it will be no longer needed to do `_determine` inside ledctl.

Reply:
Removed `led_off_all` from public API, now only used internally by `ledctl`.